src/Microsoft.OData.Edm/Vocabularies/VocabularyModelProvider.cs

```diff
@@ -0,0 +1,120 @@
+﻿//---------------------------------------------------------------------
+// <copyright file="VocabularyModelProvider.cs" company="Microsoft">
+//      Copyright (C) Microsoft Corporation. All rights reserved. See License.txt in the project root for license information.
+// </copyright>
+//---------------------------------------------------------------------
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Reflection;
+using System.Xml;
+using Microsoft.OData.Edm.Csdl;
+using Microsoft.OData.Edm.Validation;
+
+namespace Microsoft.OData.Edm.Vocabularies.V1
+{
+    internal static class VocabularyModelProvider
+    {
+        /// <summary>
+        /// Core Vocabulary Model.
+        /// </summary>
+        public static readonly IEdmModel CoreModel;
+
+        /// <summary>
+        /// Capabilities Vocabulary Model.
+        /// </summary>
+        public static readonly IEdmModel CapabilitesModel;
+
+        /// <summary>
+        /// Alternate Vocabulary Model.
+        /// </summary>
+        public static readonly IEdmModel AlternateKeyModel;
+
+        /// <summary>
+        /// Community Vocabulary Model.
+        /// </summary>
+        public static readonly IEdmModel CommunityModel;
+
+        /// <summary>
+        /// Validation Vocabulary Model.
+        /// </summary>
+        public static readonly IEdmModel ValidationModel;
+
+        /// <summary>
+        /// Authorization Vocabulary Model.
+        /// </summary>
+        public static readonly IEdmModel AuthorizationModel;
+
+        /// <summary>
+        /// All Vocabulary Models.
+        /// </summary>
+        public static readonly IEnumerable<IEdmModel> VocabularyModels;
+
+        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Performance", "CA1810:InitializeReferenceTypeStaticFieldsInline")]
+        static VocabularyModelProvider()
+        {
+            Assembly assembly = typeof(VocabularyModelProvider).GetAssembly();
+
+            string[] allResources = assembly.GetManifestResourceNames();
+
+            // core
+            string coreVocabularies = allResources.FirstOrDefault(x => x.Contains("CoreVocabularies.xml"));
+            Debug.Assert(coreVocabularies != null, "CoreVocabularies.xml: not found.");
+            CoreModel = LoadSchemaEdmModel(assembly, coreVocabularies, new IEdmModel[] { });
+
+            // capabilites
+            string capabilitiesVocabularies = allResources.FirstOrDefault(x => x.Contains("CapabilitiesVocabularies.xml"));
+            Debug.Assert(capabilitiesVocabularies != null, "CapabilitiesVocabularies.xml: not found.");
+            CapabilitesModel = LoadCsdlEdmModel(assembly, capabilitiesVocabularies, new[] { CoreModel }); // capabilities relies on core

```

Comment:
>CoreModel [](start = 92, length = 9)

I like that the referenced models are included here, but please make sure that any searches across referenced models:
1) cannot get into an infinite loop
2) ideally, don't search the same vocabulary multiple times
3) follow OData semantics

re: #3), OData says that any type used within a model must be defined within that model, or a model *directly* referenced by that model.

Reply:
I double check the Extension method such as:
FindType()
FindTerm()
FindBoundOperations()
FindOperations()

are search at current model and its  directly referenced model. So,
# 1, no infinite loop
# 2, don't search the same vocabulary multiple time.
# 3, Yes.   

---
In reply to: [248428527](https://github.com/OData/odata.net/pull/1367#discussion_r248428527) [](ancestors = 248428527)

Reply:
I also added a test to test the an infinite loop

---
In reply to: [248816140](https://github.com/OData/odata.net/pull/1367#discussion_r248816140) [](ancestors = 248816140,248428527)